File: controller/dart/density_state.dart
class DensityState {
  List<int> car_density;

  List<int> prime_capacities;

  DensityState({
    this.car_density = const [],
    this.prime_capacities = const [100, 100, 100, 100],
  });

  List<int> getDenseLanePair() {
    List<int> pair = [];
    int largest_density = 0;
    if (car_density[1] + car_density[5] > largest_density) {
      largest_density = car_density[1] + car_density[5];
      pair = [1, 5];
    }
    if (car_density[3] + car_density[7] > largest_density) {
      largest_density = car_density[3] + car_density[7];
      pair = [3, 7];
    }
    if (car_density[0] + car_density[1] > largest_density) {
      largest_density = car_density[0] + car_density[1];
      pair = [0, 1];
    }
    if (car_density[2] + car_density[3] > largest_density) {
      largest_density = car_density[2] + car_density[3];
      pair = [2, 3];
    }
    if (car_density[4] + car_density[5] > largest_density) {
      largest_density = car_density[4] + car_density[5];
      pair = [4, 5];
    }
    if (car_density[6] + car_density[7] > largest_density) {
      largest_density = car_density[6] + car_density[7];
      pair = [6, 7];
    }
    if (car_density[0] + car_density[7] > largest_density) {
      largest_density = car_density[0] + car_density[7];
      pair = [0, 7];
    }
    if (car_density[2] + car_density[7] > largest_density) {
      largest_density = car_density[2] + car_density[7];
      pair = [2, 7];
    }
    if (car_density[4] + car_density[3] > largest_density) {
      largest_density = car_density[4] + car_density[3];
      pair = [4, 3];
    }
    if (car_density[6] + car_density[5] > largest_density) {
      largest_density = car_density[6] + car_density[5];
      pair = [6, 5];
    }
    if (car_density[0] + car_density[4] > largest_density) {
      largest_density = car_density[0] + car_density[4];
      pair = [0, 4];
    }
    if (car_density[2] + car_density[6] > largest_density) {
      largest_density = car_density[2] + car_density[6];
      pair = [2, 6];
    }
    return pair;
  }

  List<int> getBestLimitReachingPair(List<int> limitReachingLanes) {
    //TODO use the getLanePairs method instead of checking all
    List<int> bestPair = [];
    int largest_density = 0;
    if ((limitReachingLanes.contains(1) && limitReachingLanes.contains(5)) &&
        (car_density[1] + car_density[5] > largest_density)) {
      largest_density = car_density[1] + car_density[5];
      bestPair = [1, 5];
    }
    if ((limitReachingLanes.contains(3) && limitReachingLanes.contains(7)) &&
        (car_density[3] + car_density[7] > largest_density)) {
      largest_density = car_density[3] + car_density[7];
      bestPair = [3, 7];
    }
    if ((limitReachingLanes.contains(0) && limitReachingLanes.contains(1)) &&
        (car_density[0] + car_density[1] > largest_density)) {
      largest_density = car_density[0] + car_density[1];
      bestPair = [0, 1];
    }
    if ((limitReachingLanes.contains(2) && limitReachingLanes.contains(3)) &&
        (car_density[2] + car_density[3] > largest_density)) {
      largest_density = car_density[2] + car_density[3];
      bestPair = [2, 3];
    }
    if ((limitReachingLanes.contains(4) && limitReachingLanes.contains(5)) &&
        (car_density[4] + car_density[5] > largest_density)) {
      largest_density = car_density[4] + car_density[5];
      bestPair = [4, 5];
    }
    if ((limitReachingLanes.contains(6) && limitReachingLanes.contains(7)) &&
        (car_density[6] + car_density[7] > largest_density)) {
      largest_density = car_density[6] + car_density[7];
      bestPair = [6, 7];
    }
    if ((limitReachingLanes.contains(0) && limitReachingLanes.contains(7)) &&
        (car_density[0] + car_density[7] > largest_density)) {
      largest_density = car_density[0] + car_density[7];
      bestPair = [0, 7];
    }
    if ((limitReachingLanes.contains(2) && limitReachingLanes.contains(7)) &&
        (car_density[2] + car_density[7] > largest_density)) {
      largest_density = car_density[2] + car_density[7];
      bestPair = [2, 7];
    }
    if ((limitReachingLanes.contains(4) && limitReachingLanes.contains(3)) &&
        (car_density[4] + car_density[3] > largest_density)) {
      largest_density = car_density[4] + car_density[3];
      bestPair = [4, 3];
    }
    if ((limitReachingLanes.contains(6) && limitReachingLanes.contains(5)) &&
        (car_density[6] + car_density[5] > largest_density)) {
      largest_density = car_density[6] + car_density[5];
      bestPair = [6, 5];
    }
    if ((limitReachingLanes.contains(0) && limitReachingLanes.contains(4)) &&
        (car_density[0] + car_density[4] > largest_density)) {
      largest_density = car_density[0] + car_density[4];
      bestPair = [0, 4];
    }
    if ((limitReachingLanes.contains(2) && limitReachingLanes.contains(6)) &&
        (car_density[2] + car_density[6] > largest_density)) {
      largest_density = car_density[2] + car_density[6];
      bestPair = [2, 6];
    }
    return bestPair;
  }

  String carDensityString() {
    String label = '';
    String density = '';
    for (int i = 0; i < 8; i++) {
      label += '${mapIndexToLabel(i)}   ';
      density += '${car_density[i]}   ';
    }
    return label + '\n' + density;
  }

  String mapIndexToLabel(int index) {
    if (index == 0)
      return "a1";
    else if (index == 1)
      return "a2";
    else if (index == 2)
      return "b1";
    else if (index == 3)
      return "b2";
    else if (index == 4)
      return "c1";
    else if (index == 5)
      return "c2";
    else if (index == 6)
      return "d1";
    else if (index == 7)
      return "d2";
    else
      throw Exception("Invalid lane index");
  }
}

File: controller/src/scheduler/density_state.py
from src.http.loopback_client import LaneData

possible_lane_pairs = [
    [1, 5],
    [3, 7],
    [0, 1],
    [2, 3],
    [4, 5],
    [6, 7],
    [0, 7],
    [2, 7],
    [4, 3],
    [6, 5],
    [0, 4],
    [2, 6],
]


class DensityState(object):

    def __init__(self,
                 car_density=None):
        self.car_density: list[int] = [] if car_density is None else car_density

    def get_dense_lane_pair(self) -> [int]:
        dense_pair: [int] = []
        largest_density: int = 0

        for pair in possible_lane_pairs:
            if self.car_density[pair[0]] + self.car_density[pair[1]] > largest_density:
                largest_density = self.car_density[pair[0]] + self.car_density[pair[1]]
                dense_pair = pair
        return dense_pair

    def get_best_limit_reaching_pair(self, limit_reaching_lanes: [int]):
        best_pair: [int] = []
        largest_density: int = 0

        for pair in possible_lane_pairs:
            if ((pair[0] in limit_reaching_lanes and pair[1] in limit_reaching_lanes) and
                    (self.car_density[pair[0]] + self.car_density[pair[1]] > largest_density)):
                largest_density = self.car_density[pair[0]] + self.car_density[pair[1]]
                best_pair = pair
        return best_pair

    def car_density_string(self) -> str:
        label: str = ''
        density: str = ''
        for i in range(0, 8):
            label += DensityState.map_index_to_label(i) + '   '
            density += str(self.car_density[i]) + '   '
        return label + '\n' + density

    def get_lane_data(self, lane_index: int) -> LaneData:
        if lane_index == 0:
            return LaneData(side="a", lane_number=1, vehicle_count=self.car_density[lane_index])
        elif lane_index == 1:
            return LaneData(side="a", lane_number=2, vehicle_count=self.car_density[lane_index])
        elif lane_index == 2:
            return LaneData(side="b", lane_number=1, vehicle_count=self.car_density[lane_index])
        elif lane_index == 3:
            return LaneData(side="b", lane_number=2, vehicle_count=self.car_density[lane_index])
        elif lane_index == 4:
            return LaneData(side="c", lane_number=1, vehicle_count=self.car_density[lane_index])
        elif lane_index == 5:
            return LaneData(side="c", lane_number=2, vehicle_count=self.car_density[lane_index])
        elif lane_index == 6:
            return LaneData(side="d", lane_number=1, vehicle_count=self.car_density[lane_index])
        elif lane_index == 7:
            return LaneData(side="d", lane_number=2, vehicle_count=self.car_density[lane_index])
        else:
            raise Exception("Invalid lane index")

    @staticmethod
    def map_index_to_label(index: int) -> str:
        if index == 0:
            return "a1"
        elif index == 1:
            return "a2"
        elif index == 2:
            return "b1"
        elif index == 3:
            return "b2"
        elif index == 4:
            return "c1"
        elif index == 5:
            return "c2"
        elif index == 6:
            return "d1"
        elif index == 7:
            return "d2"
        else:
            raise Exception("Invalid lane index")


